bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthConnector.java

```diff
@@ -0,0 +1,341 @@
+            Request request = getMethod(httpClient, tokenUrl);
+            Fields fields = initFields(GRANT_TYPE, PASSWORD, USERNAME, username, PASSWORD, password, SCOPE, scope);
+
+            setAuthentication(clientId, clientSecret, request, fields, supportsBasicAuth);
+            return doRequest(PASSWORD, httpClient, request, fields);
+        } finally {
+            shutdownQuietly(httpClient);
+        }
+    }
+
+    /**
+     * Refresh Token
+     *
+     * @see https://tools.ietf.org/html/rfc6749#section-6
+     *
+     * @param tokenUrl
+     * @param refreshToken
+     * @param clientId
+     * @param clientSecret
+     * @param scope
+     * @param supportsBasicAuth
+     * @return
+     * @throws OAuthError
+     * @throws OAuthException
+     */
+    public AccessTokenResponse grantTypeRefreshToken(@NonNull String tokenUrl, @NonNull String refreshToken,
+            @Nullable String clientId, @Nullable String clientSecret, @Nullable String scope, boolean supportsBasicAuth)
+            throws OAuthError, OAuthException {
+
+        HttpClient httpClient = null;
+        try {
+            httpClient = createHttpClient(tokenUrl);
+            Request request = getMethod(httpClient, tokenUrl);
+            Fields fields = initFields(GRANT_TYPE, REFRESH_TOKEN, REFRESH_TOKEN, refreshToken, SCOPE, scope);
+
+            setAuthentication(clientId, clientSecret, request, fields, supportsBasicAuth);
+            return doRequest(REFRESH_TOKEN, httpClient, request, fields);
+        } finally {
+            shutdownQuietly(httpClient);
+        }
+    }
+
+    /**
+     * Authorization Code Grant - part (E)
+     *
+     * @see https://tools.ietf.org/html/rfc6749#section-4.1.3
+     *
+     * @param tokenUrl
+     * @param authorizationCode
+     * @param clientId
+     * @param clientSecret
+     * @param redirectUrl
+     * @param supportsBasicAuth
+     * @return
+     * @throws OAuthError
+     * @throws OAuthException
+     */
+    public AccessTokenResponse grantTypeAuthorizationCode(@NonNull String tokenUrl, @NonNull String authorizationCode,
+            @NonNull String clientId, @Nullable String clientSecret, @NonNull String redirectUrl,
+            boolean supportsBasicAuth) throws OAuthError, OAuthException {
+        HttpClient httpClient = null;
+        try {
+            httpClient = createHttpClient(tokenUrl);
+            Request request = getMethod(httpClient, tokenUrl);
+            Fields fields = initFields(GRANT_TYPE, AUTHORIZATION_CODE, CODE, authorizationCode, REDIRECT_URI,
+                    redirectUrl);
+
+            setAuthentication(clientId, clientSecret, request, fields, supportsBasicAuth);
+            return doRequest(AUTHORIZATION_CODE, httpClient, request, fields);
+        } finally {
+            shutdownQuietly(httpClient);
+        }
+    }
+
+    /**
+     * Client Credentials Grant
+     *
+     * @see https://tools.ietf.org/html/rfc6749#section-4.4
+     *
+     * @param tokenUrl
```

Comment:
if you feel param or exception descriptions would be redundant, you can remove the following lines